docs/02-app/01-building-your-application/09-authentication/index.mdx

```diff
@@ -0,0 +1,345 @@
+
+In the example below, we check the user's role before allowing the action to proceed:
+
+```ts filename="app/lib/actions.ts" switcher
+'use server'
+
+// ...
+
+export async function serverAction(request) {
+  const session = await getSession()
+  const userRole = session?.user?.role
+
+  // Check if user is authorized to perform the action
+  if (userRole !== 'admin') {
+    throw new Error('Unauthorized access: User does not have admin privileges.')
+  }
+
+  // Proceed with the action for authorized users
+  // ... implementation of the action
+}
+```
+
+### Protecting Route Handlers
+
+Route Handlers in Next.js play a vital role in managing incoming requests. Just like server actions, they should be secured to ensure that only authorized users can access certain functionalities. This often involves verifying the user's authentication status and their permissions.
+
+Here's an example of securing a Route Handler:
+
+```ts filename="app/api/route.ts" switcher
+export async function GET() {
+  // User authentication and role verification
+  const session = await getSession()
+
+  // Check if the user is authenticated
+  if (!session) {
+    return new Response(null, { status: 401 }) // User is not authenticated
+  }
+
+  // Check if the user has the 'admin' role
+  if (session.user.role !== 'admin') {
+    return new Response(null, { status: 403 }) // User is authenticated but does not have the right permissions
+  }
+
+  // Data fetching for authorized users
+  // {...}
+}
+```
+
+This example demonstrates a Route Handler with a two-tier security check for authentication and authorization. It first checks for an active session, and then verifies if the logged-in user is an 'admin'. This approach ensures secure access, limited to authenticated and authorized users, maintaining robust security for request processing.
+
+### Authorization Using Server Components
+
+[Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) in Next.js are designed for server-side execution and offer a secure environment for integrating complex logic like authorization. They enable direct access to back-end resources, optimizing performance for data-heavy tasks and enhancing security for sensitive operations.
+
+**Key Advantages of Server Components:**
+
+- Server-Side Execution: Offloads complex tasks from the client, improving performance.
+- Direct Back-End Access: Simplifies interactions with server resources, reducing the need for extra API endpoints.
+- Enhanced Security: Ideal for critical operations like authorization, as server-side execution minimizes exposure to vulnerabilities.
```

Comment:
You could remove _the advantages using Server Components_ as it's a tangent to auth. It should be ok as the benefits are outlined in the Server Components docs you linked above.